gluten-core/src/main/scala/io/glutenproject/expression/ExpressionTransformer.scala

```diff
@@ -18,6 +18,75 @@ package io.glutenproject.expression
 
 import io.glutenproject.substrait.expression.ExpressionNode
 
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.optimizer.NormalizeNaNAndZero
+
 trait ExpressionTransformer {
   def doTransform(args: java.lang.Object): ExpressionNode
 }
+
+object LeafExpressionTransformer {
+  def apply(substraitExprName: String, original: Expression): ExpressionTransformer = {
+    new NaryExpressionTransformer(substraitExprName, Seq(), original)
+  }
+}
+
+object UnaryExpressionTransformer {
+  def apply(
+      substraitExprName: String,
+      child: ExpressionTransformer,
+      original: Expression): ExpressionTransformer = {
+    original match {
+      case c: CheckOverflow =>
+        new CheckOverflowTransformer(substraitExprName, child, c)
+      case m: MakeDecimal =>
+        new MakeDecimalTransformer(substraitExprName, child, m)
+      case k: KnownFloatingPointNormalized =>
+        new KnownFloatingPointNormalizedTransformer(child, k)
+      case n: NormalizeNaNAndZero =>
+        new NormalizeNaNAndZeroTransformer(child, n)
+      case p: PromotePrecision =>
+        new PromotePrecisionTransformer(child, p)
+      case _: GetDateField | _: GetTimeField =>
+        new ExtractDateTransformer(substraitExprName, child, original)
+      case _ =>
+        new NaryExpressionTransformer(substraitExprName, Seq(child), original)
+    }
+  }
+}
+
+object BinaryExpressionTransformer {
+  def apply(
+      substraitExprName: String,
+      left: ExpressionTransformer,
+      right: ExpressionTransformer,
+      original: Expression): ExpressionTransformer = {
+    original match {
+      case _: Like => new LikeTransformer(substraitExprName, left, right, original)
+      case _ => new NaryExpressionTransformer(substraitExprName, Seq(left, right), original)
+    }
+  }
+}
+
+object TernaryExpressionTransformer {
+  def apply(
+      substraitExprName: String,
+      first: ExpressionTransformer,
+      second: ExpressionTransformer,
+      third: ExpressionTransformer,
+      original: Expression): ExpressionTransformer = {
+    new NaryExpressionTransformer(substraitExprName, Seq(first, second, third), original)
```

Comment:
If terneary transformer just reuses nary transformer, why do we still need a ternary transformer?

Reply:
I also want to remove ternary transformer completely, do you think it's necessary to retain `UnaryExpressionTransformer` and `BinaryExpressionTransformer`? They contain some case clauses.

Reply:
I see. Looks to me they can also be replaced with `NaryExpressionTransformer` but this requires more efforts. How do you think? Maybe that should belong to another PR.

Reply:
So, we completely remove these Unary/Binary/Ternary/Quaternary transformers and one thing that's not certain is whether we should move the case clauses from `UnaryExpressionTransformer ` to `ExpressionConverter` or `NaryExpressionTransformer `. Maybe for the sake of consistency, we can move them to `ExpressionConverter` first?

Reply:
I agree. In ExpressionConverter, some special cases can be handled, and others are processed in NaryExpressionTransformer.